packages/sessions-sdk/src/index.ts

```diff
@@ -1 +1,254 @@
+import {
+  createAssociatedTokenAccountIdempotentInstruction,
+  getAssociatedTokenAddress,
+} from "@solana/spl-token";
+import type { TransactionError } from "@solana/web3.js";
+import {
+  Ed25519Program,
+  Keypair,
+  PublicKey,
+  TransactionInstruction,
+} from "@solana/web3.js";
+
+import type { SessionAdapter, TransactionResult } from "./adapter.js";
+import { TransactionResultType } from "./adapter.js";
+
+export { TransactionResultType, createSolanaWalletAdapter } from "./adapter.js";
+
+// eslint-disable-next-line no-constant-binary-expression, @typescript-eslint/no-unnecessary-condition, valid-typeof
+const IS_BROWSER = typeof globalThis.window !== undefined;
+
+const MESSAGE_HEADER = `Fogo Sessions:
+Signing this intent will allow this app to interact with your on-chain balances. Please make sure you trust this app and the domain in the message matches the domain of the current web application.
+`;
+
+type EstablishSessionOptions = {
+  adapter: SessionAdapter;
+  domain?: string | undefined;
+  expires: Date;
+  tokens: Map<PublicKey, number>;
+  extra?: string | undefined;
+};
+
+export const establishSession = async (
+  options: EstablishSessionOptions,
+): Promise<EstablishSessionResult> => {
+  const sessionKey = Keypair.generate();
+
+  const tokenInfo = await getTokenInfo(options);
+
+  const [intentInstruction, startSessionInstruction] = await Promise.all([
+    buildIntentInstruction(options, sessionKey, tokenInfo),
+    buildStartSessionInstruction(options, sessionKey, tokenInfo),
+  ]);
+  const result = await options.adapter.sendTransaction(sessionKey, [
+    intentInstruction,
+    ...buildCreateAssociatedTokenAccountInstructions(options, tokenInfo),
+    startSessionInstruction,
+  ]);
+
+  switch (result.type) {
+    case TransactionResultType.Success: {
+      return EstablishSessionResult.Success(result.signature, {
+        sessionKey,
+        publicKey: options.adapter.publicKey,
+        payer: options.adapter.payer,
+        sendTransaction: (instructions) =>
+          options.adapter.sendTransaction(sessionKey, instructions),
+      });
+    }
+    case TransactionResultType.Failed: {
+      return EstablishSessionResult.Failed(result.signature, result.error);
+    }
+  }
+};
+
+type TokenInfo = {
+  symbol: string;
+  metadataAddress: PublicKey;
+  amount: number;
+  ata: PublicKey;
+  mint: PublicKey;
+};
+
+const getTokenInfo = async (
+  options: EstablishSessionOptions,
+): Promise<TokenInfo[]> => {
+  const umi = createUmi(options.adapter.connection.rpcEndpoint);
+  return Promise.all(
+    options.tokens.entries().map(async ([mint, amount]) => {
+      const metaplexMint = metaplexPublicKey(mint.toBase58());
+      const metadataAddress = findMetadataPda(umi, { mint: metaplexMint })[0];
+      const [metadata, ata] = await Promise.all([
+        fetchMetadata(umi, metadataAddress),
+        getAssociatedTokenAddress(mint, options.adapter.publicKey),
```

Comment:
you have a sync version for this method... I think it's called `getAssociatedTokenAddress`